Doc/reference/datamodel.rst

```diff
@@ -889,57 +891,241 @@ Attribute assignment updates the module's namespace dictionary, e.g.,
+   is set to a :class:`str`.
+
+   It is **strongly** recommended that you use
+   :attr:`module.__spec__.package <importlib.machinery.ModuleSpec.parent>`
+   instead of :attr:`!module.__package__`. :attr:`__package__` is now only used
+   as a fallback if :attr:`!__spec__.parent` is not set, and this fallback
+   path is deprecated.
+
+   .. versionchanged:: 3.4
+      This attribute now defaults to ``None`` for modules created dynamically
+      using the :class:`types.ModuleType` constructor.
+      Previously the attribute was optional.
+
+   .. versionchanged:: 3.6
+      The value of :attr:`!__package__` is expected to be the same as
+      :attr:`__spec__.parent <importlib.machinery.ModuleSpec.parent>`.
+      :attr:`__package__` is now only used as a fallback during import
+      resolution if :attr:`!__spec__.parent` is not defined.
+
+   .. versionchanged:: 3.10
+      :exc:`ImportWarning` is raised if an import resolution falls back to
+      :attr:`!__package__` instead of
+      :attr:`__spec__.parent <importlib.machinery.ModuleSpec.parent>`.
+
+   .. versionchanged:: 3.12
+      Raise :exc:`DeprecationWarning` instead of :exc:`ImportWarning` when
+      falling back to :attr:`!__package__` during import resolution.
 
-      .. versionchanged:: 3.14
-         Annotations are now :ref:`lazily evaluated <lazy-evaluation>`.
-         See :pep:`649`.
+   .. deprecated-removed:: 3.13 3.15
+      :attr:`!__package__` will cease to be set or taken into consideration
+      by the import system or standard library.
 
-   :attr:`~object.__annotate__`
-      The :term:`annotate function` for this module, or ``None``
-      if the module has no annotations. See :attr:`object.__annotate__`.
+.. attribute:: module.__loader__
 
-      .. versionadded:: 3.14
+   The :term:`loader` object that the import machinery used to load the module.
+
+   This attribute is mostly useful for introspection, but can be used for
+   additional loader-specific functionality, for example getting data
+   associated with a loader.
+
+   :attr:`!__loader__` defaults to ``None`` for modules created dynamically
+   using the :class:`types.ModuleType` constructor;
+   use :func:`importlib.util.module_from_spec` instead to ensure the attribute
+   is set to a :term:`loader` object.
+
+   It is **strongly** recommended that you use
+   :attr:`module.__spec__.loader <importlib.machinery.ModuleSpec.loader>`
+   instead of :attr:`!module.__loader__`.
+
+   .. versionchanged:: 3.4
+      This attribute now defaults to ``None`` for modules created dynamically
+      using the :class:`types.ModuleType` constructor.
+      Previously the attribute was optional.
+
+   .. deprecated-removed:: 3.12 3.14
+      The value of :attr:`!module__loader__` is expected to be the same as
+      :attr:`__spec__.loader <importlib.machinery.ModuleSpec.loader>`.
+      The use of :attr:`!module.__loader__` is deprecated and slated for
+      removal in Python 3.14.
```

Comment:
At @warsaw, this deprecation note (more-or-less copied from the existing reference for this attribute) feels a lot less precise than some of the others in this section. What exactly is deprecated -- reading the attribute? Writing the attribute? Setting the attribute without setting `module.__spec__.loader`, such that Python's import machinery must fallback to `module.__loader__` rather than using `module.__spec__.loader`?

Reply:
Specifically what's deprecated is the import machinery (and other bits of Python) falling back to `module.__loader__`, but also anything in Python setting that attribute.   Eventually we plan on ripping out all mention of `module.__loader__`.

CC: @brettcannon

Reply:
Fab! I took a stab at improving it a bit in https://github.com/python/cpython/pull/124709/commits/fd67e0af1a0d9d16ccd3d9c846310acfeea2fc1f.

Should there also be a deprecation warning in the docs for `__path__`? There isn't one in the docs currently: https://docs.python.org/3/reference/import.html#path__